python/triton/experimental/gluon/language/amd/cdna4/async_copy.py

```diff
@@ -0,0 +1,159 @@
+
+
+@builtin
+def buffer_load_to_shared(dest, ptr, offsets, mask=None, other=None, cache_modifier="", _semantic=None):
+    """
+    AMD buffer load to shared operation. Buffer load is similar to global load
+    but it accesses global memory via a scalar base pointer and a tensor of
+    32-bit offsets instead of a tensor of pointers. This operation loads data
+    directly from global memory to shared memory without going through
+    registers. It happens asynchronously and requires a subsequent `async_wait`
+    to ensure the data is available in shared memory.
+
+    While using this operation, the following conditions must be met or
+    lowering to LLVM will fail:
+
+    - For the `offsets` layout, size per thread * bits per element must be 128 or 32
+    - Writes to `dest` must be coalesced.
+    - If `dest` is swizzled, it only can be swizzled within warp boundary.
+
+    Args:
+        dest (shared_memory_descriptor): Destination shared memory descriptor.
+        ptr (pointer to scalar): Global memory scalar base pointer to load from.
+        offsets (tensor): Offsets tensor for the load operation.
+        mask (tensor, optional): Mask tensor for predicated loads. Defaults to None.
+        other (tensor, optional): Tensor providing default values for masked elements. Defaults to None.
+        cache_modifier (str): Cache modifier specifier. Defaults to "".
+    """
+    _check(isinstance(offsets.type.layout, (BlockedLayout, SliceLayout)),
+           lambda: "expected offsets type layout to be BlockedLayout or SliceLayout")
+
+    mask = _unwrap_if_constexpr(mask)
+    if mask is not None:
+        offsets, mask = _semantic.broadcast_impl_value(offsets, mask)
+    other = _unwrap_if_constexpr(other)
+    if other is not None:
+        offsets, other = _semantic.broadcast_impl_value(offsets, other)
+
+    _verify_buffer_load_store(ptr, offsets, mask, other)
+
+    mask = mask.handle if mask is not None else ir.value()
+    other = other.handle if other is not None else ir.value()
+    stride = ir.value()
+    cache_modifier = _semantic._str_to_load_cache_modifier(cache_modifier)
+
+    _semantic.builder.create_buffer_load_to_local(dest.handle, ptr.handle, offsets.handle, mask, other, stride,
+                                                  cache_modifier)
+
+
+@builtin
+def async_wait(num_outstanding=0, _semantic=None):
+    """
+    Wait for outstanding asynchronous memory operations, including
+    `global_load_to_shared` and `buffer_load_to_shared`. It will block until
+    the number of outstanding operations is less than or equal to `num_outstanding`.
+
+    Args:
+        num_outstanding (int): The number of outstanding operations to wait for. Defaults to 0.
+    """
+    num_outstanding = _unwrap_if_constexpr(num_outstanding)
+    _semantic.builder.create_async_wait_group(num_outstanding)
+
+
+@builtin
+def relax_shared(smem, _semantic=None):
```

Comment:
Bikeshedding: I feel this name is a bit too general. Any other better names to reflect what's it's for so easier for kernel developer to grasp and use? Something like "async_hint_shared" etc?

Reply:
I intentionally used a more general name here because adding a hint for LLVM to avoid emitting `waitcnt` is technically a workaround in our current implementation. Ideally LLVM should figure out it based on manually inserted `waitcnt` and data flow. So probably we should not reveal these details to user?

Reply:
Good thinking! I'm not sure we can totally hide it from the developer given the low-level nature of Gluon. The root issue is that LLVM is single CPU thread (that is, single GPU warp in our context) and it's not that suitable/capable to figure out how multi threads (multiple GPU warps) async behavior, without frontend help. So I think we need to expose this. I generally prefer APIs to be self documenting so that it's clear what's it's about from reading the code directly.